kafka-client/src/main/java/dev/responsive/utils/StoreUtil.java

```diff
@@ -65,4 +81,73 @@ public static long durationToMillis(final Duration duration, final String parame
       throw new IllegalArgumentException(errorMsgPrefix + " due to arithmetic exception", e);
     }
   }
+
+  private StoreUtil() {
+  }
+
+  private static Optional<Boolean> isCompacted(final String topicName, final Admin admin) {
+    final ConfigResource resource = new ConfigResource(ConfigResource.Type.TOPIC, topicName);
+    final List<ConfigResource> request = List.of(resource);
+    final DescribeConfigsResult result = admin.describeConfigs(request);
+    final Map<ConfigResource, Config> topicConfigs;
+    try {
+      topicConfigs = result.all().get();
+    } catch (final InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (!topicConfigs.containsKey(resource)) {
+      return Optional.empty();
+    }
+    final Config topicConfig = topicConfigs.get(resource);
+    final ConfigEntry cleanupPolicyEntry = topicConfig.get(TopicConfig.CLEANUP_POLICY_CONFIG);
+    boolean compaction = cleanupPolicyEntry != null
+        && cleanupPolicyEntry.value().contains(TopicConfig.CLEANUP_POLICY_COMPACT);
+    return Optional.of(compaction);
+  }
+
+  private static boolean changelogIsSource(
+      final TopologyDescription description,
+      final String topicName
+  ) {
+    for (final var subtopology : description.subtopologies()) {
+      for (final Node node : subtopology.nodes()) {
+        if (node instanceof Source) {
+          // We ignore the topic regex field here because tables can only source a single topic
+          if (((Source) node).topicSet().contains(topicName)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  public static boolean shouldTruncateChangelog(
+      final String changelogTopic,
+      final Admin admin,
+      final Map<String, Object> appConfigs
+  ) {
+    // TODO: this is doing a lot of admin calls - we should cache the result
+    final Optional<Boolean> compacted = isCompacted(changelogTopic, admin);
+    final boolean source = changelogIsSource(
+        InternalConfigs.loadTopologyFromConfig(appConfigs),
+        changelogTopic
+    );
+    LOG.info("Changelog topic {}: compacted({}), source({})", changelogTopic, compacted, source);
+    if (compacted.isEmpty()) {
+      LOG.warn("Could not find topic {}. This should not happen. Will not truncate.",
+          changelogTopic
+      );
+      return false;
+    }
+    if (compacted.get()) {
```

Comment:
if it's compact & delete do we want to truncate?

Reply:
I'm not sure. I conservatively assumed Kafka would fail the delete whenever there was compaction enabled.

Reply:
I suppose we should test this. I did try asking in the Confluent Community slack just now, so we'll see. I'm not sure how active the other channels are

Reply:
Based on the discussion (thanks for following up on it @ableegoldman !) from community slack (https://confluentcommunity.slack.com/archives/C499EFQS0/p1688760013869749) - it should be safe to truncate `compact, delete` topics. I'll update.